### PR TITLE
Adds Support for AWS Custom Endpoints

### PR DESCRIPTION
**Pull request by danehans**
Adds support for using custom endpoints with AWS DNS provider clients.

Supports [NE-350](https://issues.redhat.com/browse/NE-350)

/assign @Miciah 
/cc @frobware @sgreene570

### CODE REVIEW

File: pkg/operator/controller/infrastructure/controller.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	}
+	if !cmp.Equal(a, b, conditionCmpOpts...) {
+		return false
+	}
+	return true
+}
+
+// ensureDNSClients...
+func (r *reconciler) ensureDNSClients() error {
+	// TODO
```
Comment (Miciah) [inline pkg/operator/controller/infrastructure/controller.go:96]: The DNS controller holds a reference to the provider object, and that reference needs to be updated, so my thinking was that the DNS controller should watch the cluster config and handle updates.  What do you think of that approach?  Proof of concept:  https://github.com/openshift/cluster-ingress-operator/pull/417

**Comment by danehans** [comment]
```
error: some steps failed:
  * could not run steps: step verify failed: test "verify" failed: the pod ci-op-ndp3dbsc/verify failed after 15s (failed containers: test): ContainerFailed one or more containers exited
Container test exited with code 2, reason Error
---
ools v0.0.0-20200115044656-831fdb1e1868
```

/test verify
/test e2e-aws-operator
/test e2e-aws

File: pkg/util/network.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	SchemeHTTPS = "https"
+)
+
+// URI validates uri as being a http(s) valid url and returns the url scheme.
+func URI(uri string) (string, error) {
+	parsed, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return "", err
+	}
+	if !parsed.IsAbs() {
```
[the remark this answers is not in the document]
Comment (danehans) [reply pkg/util/network.go:21]: Thanks for the review! From the func [godocs](https://pkg.go.dev/net/url?tab=doc#ParseRequestURI):

"It assumes that rawurl was received in an HTTP request, so the rawurl is interpreted only as an absolute URI or an absolute path."

An absolute URI must contain a scheme. See [this example](https://play.golang.org/p/Hy_PBVwAu7s).

File: pkg/util/network.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		return "", fmt.Errorf("failed validating URI, no scheme for URI %q", uri)
+	}
+	if port := parsed.Port(); len(port) != 0 {
+		intPort, err := strconv.Atoi(port)
+		if err != nil {
+			return "", fmt.Errorf("failed converting port to integer for URI %q: %v", uri, err)
```
[the remark this answers is not in the document]
Comment (danehans) [reply pkg/util/network.go:27]: `url.ParseRequestURI` does not ensure the port number is within the [valid range](https://www.ietf.org/rfc/rfc1700.txt), for [example](https://play.golang.org/p/akVmY2lu9Up). Commit `f504993` adds test cases for the lower and upper bounds of the valid port range.

File: pkg/util/network.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgreene570) [reply pkg/util/network.go:21]: Gotcha. So, if `ParseRequestURI` _must_ interpret a URI as an absolute URI (or absolute path, but that's not relevant here), and an absolute URI must contain a scheme, than isn't line 21 redundant? Looks like `ParseRequestURI` will return an error if the URI is missing a scheme (source code [here](https://github.com/golang/go/blob/go1.14.4/src/net/url/url.go#L546)).

**Comment by danehans** [comment]
`/e2e-aws-operator ` passes when run locally:
```
$ make test-e2e
GO111MODULE=on GOFLAGS=-mod=vendor go test -count 1 -v -tags e2e -run ".*" ./test/e2e
<SNIP>
    TestIngressControllerCustomEndpoints: operator_test.go:1363: deleted ingresscontroller test-custom-endpoints
--- PASS: TestIngressControllerCustomEndpoints (72.49s)
PASS
ok  	github.com/openshift/cluster-ingress-operator/test/e2e	481.002s
```

/test e2e-aws-operator

File: pkg/util/network.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danehans) [reply pkg/util/network.go]: @sgreene570 good point. Commit `a51e6ae` removes the call to `IsAbs()`.

File: go.mod [hunk 4]
```diff
@@ -13,17 +13,17 @@ require (
 	github.com/ghodss/yaml v1.0.0
 	github.com/go-logr/logr v0.1.0
 	github.com/go-logr/zapr v0.1.1
-	github.com/google/go-cmp v0.3.1
+	github.com/google/go-cmp v0.4.0
 	github.com/kevinburke/go-bindata v3.11.0+incompatible
+	github.com/onsi/gomega v1.10.1
```
Comment (sgreene570) [inline go.mod:18]: I see this PR adds the `gomega` package for `pkg/util/network_test.go`. This adds ~4Mb to the `vendor/` folder.
Is there a compelling reason to pull in `gomega`? I'm not opposed to having it, but am just wondering if there's any plans to use it more within unit tests, or something along those lines.

Comment (danehans) [reply go.mod:18]: @sgreene570 this was on my mind, so thanks for bringing it up. @Miciah thoughts ^

Comment (danehans) [reply go.mod]: @sgreene570 commit `3f1bfad` removes `gomega` in favor of table tests.

**Comment by danehans** [comment]
[deleteIngressController()](https://github.com/openshift/cluster-ingress-operator/blob/master/test/e2e/operator_test.go#L1291-L1299) is throwing an exception:

```
	/go/src/github.com/openshift/cluster-ingress-operator/test/e2e/operator_test.go:1373 +0x1f3
github.com/openshift/cluster-ingress-operator/test/e2e.assertIngressControllerDeleted(0xc0008ee500, 0x1779160, 0xc0003e1800, 0xc0001b72c0)
	/go/src/github.com/openshift/cluster-ingress-operator/test/e2e/operator_test.go:1360 +0x5b
github.com/openshift/cluster-ingress-operator/test/e2e.TestIngressControllerCustomEndpoints(0xc0008ee500)
	/go/src/github.com/openshift/cluster-ingress-operator/test/e2e/operator_test.go:1158 +0x712
```

This error is not observed when e2e is run locally:
```
$ make test-e2e
GO111MODULE=on GOFLAGS=-mod=vendor go test -count 1 -v -tags e2e -run ".*" ./test/e2e
I0721 14:49:19.893562    9666 request.go:621] Throttling request took 1.010669975s, request: GET:https://api.dhansen.devcluster.openshift.com:6443/apis/oauth.openshift.io/v1?timeout=32s
=== RUN   TestCreateIngressControllerThenSecret
    TestCreateIngressControllerThenSecret: operator_test.go:1363: deleted ingresscontroller test-9bgnq
<SNIP>
    TestIngressControllerCustomEndpoints: operator_test.go:1363: deleted ingresscontroller test-custom-endpoints
--- PASS: TestIngressControllerCustomEndpoints (32.47s)
PASS
ok  	github.com/openshift/cluster-ingress-operator/test/e2e	423.119s

**Comment by danehans** [comment]
Now it's `waitForIngressControllerCondition()` throwing an exception:
```
	/go/src/github.com/openshift/cluster-ingress-operator/test/e2e/operator_test.go:1348 +0xe5
github.com/openshift/cluster-ingress-operator/test/e2e.TestIngressControllerCustomEndpoints(0xc00089b700)
	/go/src/github.com/openshift/cluster-ingress-operator/test/e2e/operator_test.go:1155 +0x69a
```

/test e2e-aws-operator

File: test/e2e/operator_test.go [hunk 5]
```diff
@@ -1075,6 +1075,88 @@ func TestContainerLogging(t *testing.T) {
...
+		if err := kclient.Update(context.TODO(), &infraConfig); err != nil {
+			t.Logf("failed to update infrastructure config: %v\n", err)
+		}
+		// Wait for infrastructure status to update with custom endpoints.
+		err := wait.PollImmediate(1*time.Second, 30*time.Second, func() (bool, error) {
+			if err := kclient.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, &infraConfig); err != nil {
+				t.Logf("failed to get infrastructure config: %v\n", err)
+				return false, nil
+			}
+			// TODO: Change from spec to status after https://bugzilla.redhat.com/show_bug.cgi?id=1850681
```
Comment (sgreene570) [inline test/e2e/operator_test.go:1110]: The PR from BZ#1850681 has merged. 
https://github.com/openshift/cluster-config-operator/pull/143
With that in mind, should these TODO's be address in this PR?

Comment (danehans) [reply test/e2e/operator_test.go:1110]: Yeah I saw that PR merge. I plan to update this PR today.

**Comment by knobunc** [comment]
/lgtm

Let's make the CI change in a follow-up.

**Pull request merged**